File: src/main/java/org/cubexmc/listeners/CommandAllowanceListener.java
package org.cubexmc.listeners;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;

import org.cubexmc.manager.GemManager;
import org.cubexmc.manager.LanguageManager;

public class CommandAllowanceListener implements Listener {
    private final GemManager gemManager;
    private final LanguageManager languageManager;
    private final org.cubexmc.manager.ConfigManager configManager;
    private final org.cubexmc.manager.CustomCommandExecutor customCommandExecutor;

    public CommandAllowanceListener(GemManager gemManager, LanguageManager languageManager, 
                                   org.cubexmc.manager.ConfigManager configManager,
                                   org.cubexmc.manager.CustomCommandExecutor customCommandExecutor) {
        this.gemManager = gemManager;
        this.languageManager = languageManager;
        this.configManager = configManager;
        this.customCommandExecutor = customCommandExecutor;
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onPlayerCommand(PlayerCommandPreprocessEvent event) {
        String msg = event.getMessage();
        Player player = event.getPlayer();
        if (msg == null || msg.isEmpty() || player == null) return;
        if (msg.charAt(0) != '/') return;

        String raw = msg.substring(1).trim();
        if (raw.isEmpty()) return;
        String[] parts = raw.split(" ", 2);
        String label = parts[0].toLowerCase();
        java.util.UUID uid = player.getUniqueId();

        // ==================== 限次指令检查（包括扩展命令）====================
        // 优先检查是否有限次指令额度（即使玩家已有权限，限次指令也应该拦截并扣减）
        // 策略：先尝试匹配完整命令（包括子命令），如果没找到再尝试匹配主命令
        String rawLower = raw.toLowerCase(java.util.Locale.ROOT);
        boolean hasAllowance = gemManager.hasAnyAllowed(uid, rawLower);  // 先尝试完整命令
        
        if (!hasAllowance) {
            // 尝试只匹配主命令
            hasAllowance = gemManager.hasAnyAllowed(uid, label);
        }
        
        // 确定实际使用的匹配键（用于后续消耗和显示）
        String matchedLabel = hasAllowance ? 
            (gemManager.hasAnyAllowed(uid, rawLower) ? rawLower : label) : null;
        
        if (!hasAllowance) {
            // 没有限次额度：检查是否已有常规权限
            org.bukkit.command.PluginCommand pc = Bukkit.getPluginCommand(label);
            if (pc != null && pc.testPermissionSilent(player)) {
                // 有常规权限，放行
                return;
            }
            // 既没有限次也没有常规权限，不拦截（让服务器的默认权限系统处理）
            return;
        }

        // 有限次额度：检查冷却时间
        org.cubexmc.model.AllowedCommand allowedCmd = gemManager.getAllowedCommand(uid, matchedLabel);
        if (allowedCmd != null && allowedCmd.getCooldown() > 0) {
            if (!customCommandExecutor.checkCooldown(uid, matchedLabel)) {
                long remainingSeconds = customCommandExecutor.getRemainingCooldown(uid, matchedLabel);
                player.sendMessage("§c命令冷却中，还需等待 " + remainingSeconds + " 秒");
                event.setCancelled(true);
                return;
            }
        }
        
        // 尝试扣减次数
        boolean consumed = gemManager.tryConsumeAllowed(uid, matchedLabel);
        if (!consumed) {
            languageManager.sendMessage(player, "allowance.none_left");
            event.setCancelled(true);
            return;
        }

        boolean ok = false;
        
        // 检查是否为扩展命令（多命令链或带执行者前缀）
        if (allowedCmd != null && !allowedCmd.isSimpleCommand()) {
            // 扩展命令：使用 CustomCommandExecutor 执行
            String[] args = parts.length > 1 ? parts[1].split(" ") : new String[0];
            ok = customCommandExecutor.executeExtendedCommand(player, allowedCmd, args);
        } else {
            // 简单命令：玩家临时 OP 执行
            boolean wasOp = player.isOp();
            try {
                if (!wasOp) {
                    player.setOp(true);
                }
                ok = player.performCommand(raw);
            } catch (Throwable ignored) {
                ok = false;
            } finally {
                if (!wasOp && player.isOp()) {
                    player.setOp(false);
                }
            }
        }
        
        if (!ok) {
            // 执行失败，退还次数
            gemManager.refundAllowed(uid, matchedLabel);
            languageManager.sendMessage(player, "allowance.execute_failed");
            event.setCancelled(true);
            return;
        }

        // 执行成功：设置冷却时间并显示剩余次数
        if (allowedCmd != null && allowedCmd.getCooldown() > 0) {
            customCommandExecutor.setCooldown(uid, matchedLabel, allowedCmd.getCooldown());
        }
        
        int remain = gemManager.getRemainingAllowed(uid, matchedLabel);
        String remainShown = remain < 0 ? "∞" : String.valueOf(remain);
        java.util.Map<String, String> placeholders = new java.util.HashMap<>();
        placeholders.put("command", matchedLabel);
        placeholders.put("remain", remainShown);
        languageManager.sendMessage(player, "allowance.used", placeholders);
        // prevent double-execution
        event.setCancelled(true);
    }
}

File: src/main/java/org/cubexmc/listeners/CommandAllowanceListener.kt
package org.cubexmc.listeners

import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerCommandPreprocessEvent
import org.bukkit.event.player.PlayerCommandSendEvent
import org.bukkit.event.server.TabCompleteEvent
import org.cubexmc.manager.CustomCommandExecutor
import org.cubexmc.manager.GameplayConfig
import org.cubexmc.manager.GemAllowanceManager
import org.cubexmc.manager.LanguageManager
import java.util.Collections
import java.util.Locale
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class CommandAllowanceListener(
    private val allowanceManager: GemAllowanceManager,
    private val languageManager: LanguageManager,
    private val customCommandExecutor: CustomCommandExecutor,
    private val gameplayConfig: GameplayConfig?,
) : Listener {
    private val proxyLabels: MutableSet<String> = ConcurrentHashMap.newKeySet()
    private val bypassPlayers: MutableSet<UUID> = ConcurrentHashMap.newKeySet()

    /**
     * Update the set of command labels that are backed by dedicated proxy commands.
     */
    fun updateProxyLabels(labels: Set<String>?) {
        proxyLabels.clear()
        if (labels != null) {
            for (label in labels) {
                if (label.isEmpty()) {
                    continue
                }
                proxyLabels.add(label.lowercase(Locale.ROOT))
            }
        }
    }

    @EventHandler(priority = EventPriority.MONITOR)
    fun onCommandSend(event: PlayerCommandSendEvent) {
        val player = event.player
        val allowed = allowanceManager.getAvailableCommandLabels(player.uniqueId)
        if (allowed.isEmpty()) {
            return
        }
        val commands = event.commands
        for (label in allowed) {
            if (label == null || label.isEmpty()) {
                continue
            }
            commands.add(label)
        }
    }

    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
    fun onTabComplete(event: TabCompleteEvent) {
        if (event.sender !is Player) {
            return
        }
        val buffer = event.buffer
        if (buffer.isEmpty() || buffer[0] != '/') {
            return
        }
        val player = event.sender as Player
        val allowed = allowanceManager.getAvailableCommandLabels(player.uniqueId)
        if (allowed.isEmpty()) {
            return
        }

        val withoutSlash = buffer.substring(1)
        val trailingSpace = buffer.endsWith(" ")
        val parts = withoutSlash.split(" ".toRegex(), -1).toTypedArray()
        if (parts.isEmpty()) {
            return
        }

        val completions = ArrayList(event.completions)
        if (parts.size == 1 && !trailingSpace) {
            val partial = parts[0].lowercase(Locale.ROOT)
            for (label in allowed) {
                if (partial.isEmpty() || label.startsWith(partial)) {
                    completions.add(label)
                }
            }
            if (completions.isNotEmpty()) {
                event.completions = distinct(completions)
            }
            return
        }

        val base = parts[0].lowercase(Locale.ROOT)
        if (!allowed.contains(base)) {
            return
        }

        if (parts.size == 1 && trailingSpace) {
            // command typed exactly, no argument suggestions available yet
            return
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = false)
    fun onPlayerCommand(event: PlayerCommandPreprocessEvent) {
        val message = event.message
        val player = event.player
        if (message.isEmpty()) {
            return
        }
        if (message[0] != '/') {
            return
        }

        val raw = message.substring(1).trim()
        if (raw.isEmpty()) {
            return
        }
        val parts = raw.split(" ".toRegex(), 2).toTypedArray()
        val label = parts[0].lowercase(Locale.ROOT)

        if (proxyLabels.contains(label)) {
            // Proxied command: let the registered command handle execution.
            return
        }

        if (bypassPlayers.remove(player.uniqueId)) {
            return
        }

        val args = if (parts.size > 1) parts[1].split(" ".toRegex()).toTypedArray() else emptyArray()
        val handled = handleAllowedCommand(player, raw, label, args, true)
        if (handled) {
            event.isCancelled = true
        }
    }

    fun handleProxyExecution(player: Player, raw: String, label: String, args: Array<String>) {
        handleAllowedCommand(player, raw, label, args, false)
    }

    fun suggestProxyTab(player: Player?, alias: String, args: Array<String>): List<String> {
        if (player == null) {
            return Collections.emptyList()
        }
        val allowed = allowanceManager.getAvailableCommandLabels(player.uniqueId)
        if (allowed.isEmpty()) {
            return Collections.emptyList()
        }
        if (args.isEmpty()) {
            return if (allowed.contains(alias)) Collections.singletonList(alias) else Collections.emptyList()
        }
        return Collections.emptyList()
    }

    private fun handleAllowedCommand(
        player: Player?,
        raw: String?,
        baseLabel: String,
        args: Array<String>,
        allowDelegate: Boolean,
    ): Boolean {
        if (player == null || raw.isNullOrEmpty()) {
            return false
        }

        val uid = player.uniqueId
        val rawLower = raw.lowercase(Locale.ROOT)
        val labelLower = baseLabel.lowercase(Locale.ROOT)

        val fullMatch = allowanceManager.hasAnyAllowed(uid, rawLower)
        val baseMatch = fullMatch || allowanceManager.hasAnyAllowed(uid, labelLower)

        if (!baseMatch) {
            if (!allowDelegate) {
                languageManager.sendMessage(player, "command.no_permission")
                return true
            }
            val pluginCommand = Bukkit.getPluginCommand(labelLower)
            if (pluginCommand != null && pluginCommand.testPermissionSilent(player)) {
                return false
            }
            return false
        }

        val matchedLabel = if (fullMatch) rawLower else labelLower
        val resolved = allowanceManager.resolveAllowedCommand(uid, matchedLabel)
        if (resolved == null) {
            languageManager.sendMessage(player, "allowance.none_left")
            return true
        }

        val allowedCommand = resolved.command
        val cooldownKey = resolved.cooldownKey
        if (allowedCommand != null && allowedCommand.cooldown > 0) {
            if (!customCommandExecutor.checkCooldown(uid, cooldownKey)) {
                val remainingSeconds = customCommandExecutor.getRemainingCooldown(uid, cooldownKey)
                val cooldownPlaceholders = HashMap<String, String>()
                cooldownPlaceholders["seconds"] = remainingSeconds.toString()
                languageManager.sendMessage(player, "allowance.cooldown", cooldownPlaceholders)
                return true
            }
        }

        val consumed = allowanceManager.tryConsumeAllowed(uid, resolved)
        if (!consumed) {
            languageManager.sendMessage(player, "allowance.none_left")
            return true
        }

        val ok = if (allowedCommand != null && !allowedCommand.isSimpleCommand()) {
            customCommandExecutor.executeExtendedCommand(player, allowedCommand, args)
        } else {
            val useOp = gameplayConfig != null && gameplayConfig.isOpEscalationAllowed
            if (useOp) {
                // OP 提权模式（管理员显式启用）
                val wasOp = player.isOp
                try {
                    if (!wasOp) {
                        player.isOp = true
                    }
                    bypassPlayers.add(uid)
                    player.performCommand(raw)
                } catch (_: Throwable) {
                    false
                } finally {
                    bypassPlayers.remove(uid)
                    if (!wasOp && player.isOp) {
                        player.isOp = false
                    }
                }
            } else {
                // 安全模式：以控制台身份在全局线程执行（Folia 安全）
                customCommandExecutor.dispatchAsConsole(raw)
            }
        }

        if (!ok) {
            allowanceManager.refundAllowed(uid, resolved)
            languageManager.sendMessage(player, "allowance.execute_failed")
            return true
        }

        if (allowedCommand != null && allowedCommand.cooldown > 0) {
            customCommandExecutor.setCooldown(uid, cooldownKey, allowedCommand.cooldown)
        }

        val remain = allowanceManager.getRemainingAllowed(uid, matchedLabel)
        val remainShown = if (remain < 0) "∞" else remain.toString()
        val placeholders = HashMap<String, String>()
        placeholders["command"] = matchedLabel
        placeholders["remain"] = remainShown
        languageManager.sendMessage(player, "allowance.used", placeholders)
        return true
    }

    private fun distinct(values: List<String>): List<String> = ArrayList(LinkedHashSet(values))
}
